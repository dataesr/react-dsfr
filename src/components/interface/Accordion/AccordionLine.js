import classnames from 'classnames';
import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { CHILDREN_TYPE, CLASS_NAME_TYPE } from '../../../utils/variables';

const AccordionLine = (props) => {
  const {
    title, children, id, className,
  } = props;
  const expandedLine = {
    false: {
      class: 'rf-collapse', stateHeight: null, aria: 'false', expanded: false,
    },
    true: {
      class: 'rf-collapse rf-collapse--expanded', stateHeight: 'none', aria: 'true', expanded: true,
    },
  };

  const [collapse, setCollapse] = useState('0px');
  const [line, setLine] = useState(expandedLine.false);

  useEffect(() => {
    const lineElement = document.getElementById(`rf-accordion-${id}`);
    setCollapse(`-${lineElement.getBoundingClientRect().height}px`);
  }, [setCollapse, id, line]);

  return (
    <li className={classnames(className)}>
      <section className="rf-accordion">
        <h3 className="rf-accordion__title">
          <button
            onClick={() => setLine(expandedLine[!line.expanded])}
            type="button"
            className="rf-accordion__btn"
            aria-controls={`rf-accordion-${id}`}
            aria-expanded={line.aria}
          >
            {title}
          </button>
        </h3>
        <div
          style={{ maxHeight: line.stateHeight, '--collapse': collapse }}
          className={line.class}
          id={`rf-accordion-${id}`}
        >
          {children}
        </div>
      </section>
    </li>
  );
};

AccordionLine.defaultProps = {
  className: '',
};

AccordionLine.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  /**
   * One of: node, arrayOf(node), string
   */
  children: CHILDREN_TYPE.isRequired,
  /**
   * One of: string, object
   */
  className: CLASS_NAME_TYPE,
};

export default AccordionLine;
