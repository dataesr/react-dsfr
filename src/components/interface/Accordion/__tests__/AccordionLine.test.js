import { useState as useStateMock } from 'react';
import Enzyme, { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import AccordionLine from '../AccordionLine';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

const initialProps = {
  title: 'title',
  id: '00',
  children: 'accordion line',
};

describe('<AccordionLine />', () => {
  let wrapper;
  const setState = jest.fn();
  beforeEach(() => {
    useStateMock.mockImplementation((init) => [init, setState]);
    wrapper = (props = {}) => shallow(
      <AccordionLine
        title={props.title}
        id={props.id}
      >
        {props.children}
      </AccordionLine>,
    );
  });

  it('renders correctly', () => {
    const component = renderer
      .create(
        <AccordionLine
          title={initialProps.title}
          onClick={initialProps.onClick}
          id={initialProps.id}
        >
          {initialProps.children}
        </AccordionLine>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('onClick AccordionLine', () => {
    const props = {
      title: 'title',
      id: '01',
      children: 'accordion line',
    };

    const component = wrapper({ ...props });
    component.find('button').simulate('click');
    expect(setState).toHaveBeenCalledTimes(1);
  });
});
