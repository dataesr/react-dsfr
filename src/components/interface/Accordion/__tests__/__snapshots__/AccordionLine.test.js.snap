// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccordionLine /> renders correctly 1`] = `
<li
  className=""
>
  <section
    className="rf-accordion"
  >
    <h3
      className="rf-accordion__title"
    >
      <button
        aria-controls="rf-accordion-00"
        aria-expanded="false"
        className="rf-accordion__btn"
        onClick={[Function]}
        type="button"
      >
        title
      </button>
    </h3>
    <div
      className="rf-collapse"
      id="rf-accordion-00"
      style={
        Object {
          "--collapse": "0px",
          "maxHeight": null,
        }
      }
    >
      accordion line
    </div>
  </section>
</li>
`;
